### PR TITLE
Minor fixes and add unit tests

### PR DESCRIPTION
**Pull request by mmanciop**

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -20,6 +20,39 @@
...
+        logger.error("Invalid port for target: %s", port)
+        return ""
+
+    if port < 0 or port > 2 ** 16 - 1:
+        logger.error("Invalid port range for target: %s", port)
+        return ""
+
+    # validate host
+    match = re.search(
+        r"^(?:(?:(?:(?:[a-zA-Z0-9][-a-zA-Z0-9]*)?[a-zA-Z0-9])[.])*(?:[a-zA-Z][-a-zA-Z0-9]*[a-zA-Z0-9]|[a-zA-Z])[.]?)$",
```
Comment (sed-i) [inline src/charm.py:45]: This regexp does not work for plain IP addresses.
Why not use something as simple as
```python
>>> from urllib.parse import urlparse
>>> urlparse('http://localhost:9090/metrics')
ParseResult(scheme='http', netloc='localhost:9090', path='/metrics', params='', query='', fragment='')
```
I'd require the user to provide a scheme, and drop it later. Also, the `/metrics` path does not need to be its own config param.

File: src/charm.py [hunk 2]
```diff
@@ -20,6 +20,39 @@
 DEFAULT_METRICS_ENDPOINT_PORT = 80
```
Comment (sed-i) [inline src/charm.py:20]: Not sure if this default is:

1. correct: I haven't seen any port 80 metrics endpoints so far.
2. needed: if port is omitted, 80 is the default anyway, no?

Would it be so bad if we require the user to provide it explicitly?

Comment (mmanciop) [reply src/charm.py:20]: There really is no standard, but `80` is the default web port :-)

**Comment by mmanciop** [comment]
@sed-i good point about IP addresses, but the target addresses do _not_ have the scheme so the commit you pushed seems to break stuff

**Comment by mmanciop** [comment]
@sed-i I deliberately fixed the tests to how the charm *must* behave, you get to fix the code :P

File: src/charm.py [hunk 3]
```diff
@@ -41,83 +65,51 @@ def __init__(self, *args):
...
-        if not self.unit.is_leader():
-            return
-
-        jobs = self._scrape_jobs()
-        if not jobs:
-            return
-
-        for relation in self.model.relations[self._prometheus_relation]:
-            relation.data[self.app]["scrape_jobs"] = json.dumps(jobs)
+        if self.unit.is_leader():
```
Comment (balbirthomas) [inline src/charm.py:68]: Why does the existing flat code flow have to be changed to this nested code flow ?

File: src/charm.py [hunk 4]
```diff
@@ -41,83 +65,51 @@ def __init__(self, *args):
...
 
-    def _scrape_jobs(self):
-        targets = self._targets()
+                self.unit.status = ActiveStatus()
 
-        jobs = (
-            [
+    def _scrape_jobs(self) -> list:
+        if targets := self._targets():
+            labels = self._labels()
```
Comment (balbirthomas) [inline src/charm.py:77]: There was good reason why labels were not always being added. It leads to uglyness such as labels `labels: {}`. Please revert this change.

File: src/charm.py [hunk 5]
```diff
@@ -41,83 +65,51 @@ def __init__(self, *args):
...
             ]
-            if targets
-            else []
-        )
-
-        labels = self._labels()
-        if jobs and labels:
-            jobs[0]["static_configs"][0]["labels"] = labels
 
-        return jobs
```
Comment (balbirthomas) [inline src/charm.py:78]: `return jobs` has more semantic information than `return []` and hence leads to more readable code.

File: src/charm.py [hunk 6]
```diff
@@ -41,83 +65,51 @@ def __init__(self, *args):
...
+        return []
 
-    def _targets(self):
-        if not (targets := self.model.config.get("targets", "")):
+    def _targets(self) -> list:
+        if not (config_targets := self.model.config.get("targets", "")):
```
Comment (balbirthomas) [inline src/charm.py:94]: If all you were trying to do here is change a variable name to your preference then perhaps a better variable name would have been `scrape_targets` because that is what they are. They certainly are not `config_targets`.

File: src/charm.py [hunk 7]
```diff
@@ -41,83 +65,51 @@ def __init__(self, *args):
...
+                invalid_targets.append(config_target)
 
         if invalid_targets:
+            logger.error("Invalid targets found: %s", invalid_targets)
             self.unit.status = BlockedStatus(f"Invalid targets : {invalid_targets}")
+            return []
 
         return targets
 
-    def _validated_address(self, address):
```
Comment (balbirthomas) [inline src/charm.py:98]: Why delete a whole function just to copy it elsewhere ?

File: src/charm.py [hunk 8]
```diff
@@ -8,7 +8,7 @@
 
 import json
 import logging
-import re
+from urllib.parse import urlparse
```
Comment (balbirthomas) [inline src/charm.py:11]: This approach has been tried and found wanting. See commit entitled "Improve validation of target address" and commit message therein.

File: tests/unit/test_charm.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        downstream_rel_id = self.harness.add_relation("metrics-endpoint", "prometheus-k8s")
+
+        self.assertEqual(
+            {},
+            self.harness.get_relation_data(downstream_rel_id, self.harness.charm.app.name),
+        )
+
+        self.assertEqual(self.harness.model.unit.status, BlockedStatus("No targets specified"))
+
+    def test_no_leader(self):
```
Comment (balbirthomas) [inline tests/unit/test_charm.py:36]: Same point regarding test name.

File: tests/unit/test_charm.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        self.harness.update_config({"targets": "foo:1234,bar:5678"})
+
+        downstream_rel_id = self.harness.add_relation("metrics-endpoint", "prometheus-k8s")
+        relation_data = self.harness.get_relation_data(
+            downstream_rel_id, self.harness.charm.app.name
+        )
+
+        self.assertEqual({}, relation_data)
+
+    def test_no_labels(self):
```
Comment (balbirthomas) [inline tests/unit/test_charm.py:49]: Same point regarding test name.

File: tests/unit/test_charm.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+                        },
+                    ],
+                }
+            ],
+            json.loads(relation_data["scrape_jobs"]),
+        )
+
+        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
+
+    def test_with_labels(self):
```
Comment (balbirthomas) [inline tests/unit/test_charm.py:79]: Same point regarding test name.

File: tests/unit/test_charm.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+                        },
+                    ],
+                }
+            ],
+            json.loads(relation_data["scrape_jobs"]),
+        )
+
+        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
+
+    def test_valid_target_without_port(self):
```
Comment (balbirthomas) [inline tests/unit/test_charm.py:116]: Same point regarding test name.

File: tests/unit/test_charm.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+                        },
+                    ],
+                }
+            ],
+            json.loads(relation_data["scrape_jobs"]),
+        )
+
+        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
+
+    def test_invalid_host_with_scheme(self):
```
Comment (balbirthomas) [inline tests/unit/test_charm.py:146]: Same point regarding test name.

File: tests/unit/test_charm.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+            downstream_rel_id, self.harness.charm.app.name
+        )
+
+        self.assertEqual([], list(relation_data.keys()))
+
+        self.assertEqual(
+            self.harness.model.unit.status, BlockedStatus("Invalid target: 'https://foo:1234'")
+        )
+
+    def test_invalid_host_with_path(self):
```
Comment (balbirthomas) [inline tests/unit/test_charm.py:163]: Same point regarding test name.

File: tests/unit/test_charm.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+class TestCharm(unittest.TestCase):
+    def setUp(self):
+        """Flake8 forces me to write meaningless docstrings."""
+        self.harness = Harness(PrometheusScrapeTargetCharm)
+        self.harness.set_model_info(name="lma", uuid="e40bf1a0-91f4-45a5-9f35-eb30fd010e4d")
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+
+    def test_no_data(self):
```
Comment (balbirthomas) [inline tests/unit/test_charm.py:23]: Test what about `no_data` ? This test name could be better worded such as `test_charm_blocks_without_any_targets` or some such thing. Such a name gives a meaningful but cryptic summary of what the test is all about with requiring the developer to read the whole test to understand what is being tested.
Note the code comments and docstrings are usually not visible when tests fail in a test runner.

**Review by balbirthomas — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 8 above)
Comment (sed-i) [reply src/charm.py:11]: Not sure what you mean. This seems to work well. Could you take a look at my recent changes?

File: src/charm.py (hunk 3 above)
Comment (sed-i) [reply src/charm.py:73]: flattened.

File: src/charm.py (hunk 6 above)
Comment (sed-i) [reply src/charm.py:99]: I think the config_ prefix is because it is taken from the config and is not yet a "confirmed" scrape target

File: src/charm.py (hunk 8 above)
Comment (balbirthomas) [reply src/charm.py:11]: Excellent work. I particularly like the RFC comment. Just re-request review when you are done.

File: tests/unit/test_charm.py (hunk 9 above)
Comment (sed-i) [reply tests/unit/test_charm.py:36]: this test is not very useful as-is because a non-leader cannot change app data anyway. Maybe replace with `self.assertRaises(...)`?

File: src/charm.py [hunk 16]
```diff
@@ -41,83 +70,51 @@ def __init__(self, *args):
...
+            static_config = {"targets": targets}
+            if labels := self._labels():
+                static_config.update(labels=labels)
 
-        jobs = (
-            [
+            return [
                 {
                     "job_name": self._job_name(),
                     "metrics_path": self.model.config["metrics-path"],
```
Comment (sed-i) [inline src/charm.py:91]: Prometheus has already got a default for the path:
```
# The HTTP resource path on which to fetch metrics from targets.
[ metrics_path: <path> | default = /metrics ]
```
So maybe we should be duplicating it (hardcoding it ourselves)

File: src/charm.py (hunk 2 above)
Comment (sed-i) [reply src/charm.py:20]: Prometheus got it covered already:
```
<host>: a valid string consisting of a hostname or IP followed by an optional port number
```
I think we better off dropping it.

File: src/charm.py (hunk 16 above)
Comment (mmanciop) [reply src/charm.py:91]: I am in favor of removing it if left default

File: src/charm.py (hunk 6 above)
Comment (balbirthomas) [reply src/charm.py:99]: Hmm. I don't see where the question of "confirmation" arises. The charm has no procedure to "confirm" the targets. It does validate them. So if you want to be more verbose `unverified_scrape_targets`.

File: src/charm.py [hunk 17]
```diff
@@ -41,83 +70,51 @@ def __init__(self, *args):
...
-            relation.data[self.app]["scrape_jobs"] = json.dumps(jobs)
+            self.unit.status = ActiveStatus()
 
-    def _scrape_jobs(self):
-        targets = self._targets()
+    def _scrape_jobs(self) -> list:
+        if targets := self._targets():
+            static_config = {"targets": targets}
+            if labels := self._labels():
+                static_config.update(labels=labels)
```
Comment (balbirthomas) [inline src/charm.py:86]: Apart from deep nesting I do not see how this address the issue of empty labels I posted in my previous review. 

```
import this
...
Flat is better than nested.
...
```

Comment (balbirthomas) [reply src/charm.py:86]: Nested code is harder to understand, once you get past metal blocks against (inverted) negative logic.

File: src/charm.py (hunk 7 above)
Comment (balbirthomas) [reply src/charm.py:98]: How does putting the method in module scope help in this case ? Is there any other object that will use this method ? Couldn't it be refactored to module scope in the future if such a use case arises ?

**Review by balbirthomas — APPROVED** [review]

File: src/charm.py (hunk 17 above)
Comment (sed-i) [reply src/charm.py:86]: With this change, no "lables: {}" is created. Tests were updated and confirm this.

File: src/charm.py (hunk 6 above)
Comment (sed-i) [reply src/charm.py:99]: Renamed.

**Review by balbirthomas — APPROVED** [review]
